### DISCUSSION TITLE
Sceneform vs Filament in 2021

### DISCUSSION THREAD

**Discussion by paulorrfilho** [google/filament#3575] [Q&A]
Hi, I am trying to port my iOS app (DrumKnee 3D) to Android and I think Filament would be a good fit for me. I need to render a few models on the screen using a PBR approach and overlay my app menus.

As I understand Filament does not provide any type of Physics simulation in terms of gravity, any alternatives for this? 

I saw a comment on Reddit saying that one should go with Sceneform for simple 3D projects however, the lack of future support scares me. Is this still a valid comment in 2021 for a starting project?

Sorry for so many questions, just trying to get my head around the current 3D rendering situation on Android.

Thank you,
Paulo

**Comment by romainguy** [comment]
Development on Sceneform has unfortunately ceased but its code is now Open Source on GitHub and can server as a good starting point in your own app if you need AR. If all you need is 3D you can use Filament by itself.

It is correct that Filament does not provide any physics simulation. It is purely a rendering engine and nothing more. You can however use any physics engine you wish, like Bullet, ODE, etc.

**Reply by paulorrfilho** [reply]
Great, that's what I thinking.
Thank you so much for the quick reply.

**Reply by romainguy** [reply]
Note though that we do have a couple of companion libraries that may be helpful. `gltfio` for instance can be used to load glTF files. `filament-utils` contains various utilities (for KTX files or camera manipulation). But we don't provide generic 3D file format loaders (you can use `assimp` for this) or image loaders (`stb_image`, etc.).

**Reply by paulorrfilho** [reply]
What is the advantage of using glTF files? Can I change the textures in real-time as I do with .obj files?
Can I play animations created in Blender using both file types? (I am thinking of this as an alternative to the physics engine)

**Reply by romainguy** [reply]
glTF is a much more complete format than OBJ. Just to name a few advantages:
- Supports animations (skinning and morphing)
- Supports PBR materials
- Supports hierarchies and transforms
- Supports lights via an extension

glTF files can be exported from Blender and our gltfio library will let you play those animations easily. You can change textures/materials in real-time just like with OBJ. In the end you just manipulate Filament objects.

**Reply by paulorrfilho** [reply]
Thank you, is there any example of how to texture a glTF mesh? All examples for Android are using an obj file.
I don't see documentation around the Filament API itself, is there any other than the one in the code itself?

**Reply by romainguy** [reply]
Filament doesn't care about OBJ vs glTF, etc. It's all the same: you have a `Renderable` (the mesh) which contains geometries (sub-meshes), and you assign `MaterialInstance`s to those geometries (a material instance is a set of parameter for a `Material`).

That said, if you use the gltfio library we provide, the meshes will automatically be textured, etc. using the materials defined in the glTF file. If you want to override that, you simply modify the material instance provided by gltfio.

**Comment by paulorrfilho** [comment]
I am glad I spoke with you about this. I think the best option for me is indeed the glTF format.

**Comment by miinso** [comment]
@paulorrfilho There is a [Java port](https://pybullet.org/Bullet/phpBB3/viewtopic.php?t=1827) of Bullet physics SDK that can be integrated to Android runtime without the burden. 

I tried to glue them together and it did kind of work, still I'm positive using JNI binding will earn you some extra performance though.

[![filament-physics-cube-instancing](https://user-images.githubusercontent.com/46559594/124067157-dcc6a700-da74-11eb-8b40-715902b62f79.gif)](https://github.com/61315/filament-physics)

In terms of gravity, the Java port implements the original methods as-is, see [src/main/java/com/bulletphysics/dynamics/DiscreteDynamicsWorld.java#L409-L419](https://github.com/bubblecloud/jbullet/blob/00b51569461391d87436369ac723e875efef0dcd/src/main/java/com/bulletphysics/dynamics/DiscreteDynamicsWorld.java#L409-L419):
```java
@Override
public void setGravity(Vector3f gravity) {
	this.gravity.set(gravity);
	for (int i = 0; i < collisionObjects.size(); i++) {
		CollisionObject colObj = collisionObjects.getQuick(i);
		RigidBody body = RigidBody.upcast(colObj);
		if (body != null) {
			body.setGravity(gravity);
		}
	}
}
```
